### DISCUSSION TITLE
Android app crashes on startup

### DISCUSSION THREAD

**Discussion by AzuxirenLeadGuy** [AvaloniaUI/Avalonia#8133] [General]
I am building the `avalonia.xplat` application for android. When I run it on the `Debug` configuration in Visual Studio, It runs fine. However, if I close the debugging session and run the app again on the android emulator, it crashes on startup. Similarly, if I build and deploy it on the `Release` configuration, the app is deployed but it crashes on startup. 

EDIT: 

This video [shows](https://user-images.githubusercontent.com/40014666/168531872-2589e621-1da2-41d8-a843-da1ae0ac4890.mp4) what I see while I run it: 

Thanks to @timunie , The exception message is known. It is as shown

```
System.InvalidOperationException: Invalid size returned for Measure.
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Controls.Embedding.EmbeddableControlRoot.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control)
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control)
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control)
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control)
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control)
   at Avalonia.Layout.LayoutManager.ExecuteMeasurePass()
   at Avalonia.Layout.LayoutManager.InnerLayoutPass()
   at Avalonia.Layout.LayoutManager.ExecuteLayoutPass()
   at Avalonia.Controls.TopLevel.HandleResized(Size clientSize, PlatformResizeReason reason)
   at Avalonia.Android.Platform.SkiaPlatform.TopLevelImpl.OnResized(Size size)
   at Avalonia.Android.AvaloniaView.ViewImpl.OnResized(Size size)
   at Avalonia.Android.Platform.SkiaPlatform.TopLevelImpl.ViewImpl.Android.Views.ISurfaceHolderCallback.SurfaceChanged(ISurfaceHolder holder, Format format, Int32 width, Int32 height)
   at Android.Views.ISurfaceHolderCallbackInvoker.n_SurfaceChanged_Landroid_view_SurfaceHolder_III(IntPtr jnienv, IntPtr native__this, IntPtr native_holder, Int32 native_format, Int32 width, Int32 height)
```

Please let me know how to solve this. Thanks in advance.

**Comment by timunie** [comment]
Hi

Without an error message it's not really possible to help you here. I think you can understand that. But I think you also don't see an error, instead your App just crashes. 

But I have good news, you can log unhandled exceptions and write them to a log file for example. https://docs.avaloniaui.net/docs/getting-started/unhandledexceptions

I hope this helps. 

Happy coding
Tim

**Reply by AzuxirenLeadGuy** [reply]
Thanks for the response. I tried implementing the logging but storage enough, the error is still not traced.

I did this implementation to log all trace information to a file as shown:

```csharp
    public class MainActivity : AvaloniaActivity<App>
    {
        protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
        {
            string filename = GetExternalFilesDir(string.Empty) + "/tracelog.txt";
            System.Diagnostics.Trace.Listeners.Add
                (
                new System.Diagnostics.TextWriterTraceListener(filename) 
                { 
                    Filter = new System.Diagnostics.EventTypeFilter(System.Diagnostics.SourceLevels.All),
                    TraceOutputOptions = (System.Diagnostics.TraceOptions)63
                }
                );
            System.Diagnostics.Trace.Listeners.Remove("Default");
            return base.CustomizeAppBuilder(builder).LogToTrace(Avalonia.Logging.LogEventLevel.Debug);
        }
    }
```

Now, when I debug the android app using Visual Studio debugger, The app runs fine on the emulator but no lines are generated in the file. When the app crashes, no error log is generated either. The file remains empty.

Notice that I have set the `TraceLevel` at `Avalonia.Logging.LogEventLevel.Debug`. If I set it up as `Avalonia.Logging.LogEventLevel.Verbose`, there is a lot of lines generated (So I can assure that the logging action is performed correctly.
[This is the tracelog file when the program crashes.](https://github.com/AvaloniaUI/Avalonia/files/8697595/tracelog-failed.txt)
[This is the tracelog file when the program is running properly](https://github.com/AvaloniaUI/Avalonia/files/8697596/tracelog-success.txt)

I have also uploaded a video to show how it ouccers. 

Please advise what to do next.

https://user-images.githubusercontent.com/40014666/168531872-2589e621-1da2-41d8-a843-da1ae0ac4890.mp4

**Reply by timunie** [reply]
Can you show me how you are trying to catch the exception and write it to the log file? 
https://docs.avaloniaui.net/docs/getting-started/unhandledexceptions#the-global-try-catch

**Reply by AzuxirenLeadGuy** [reply]
> Can you show me how you are trying to catch the exception and write it to the log file?

I did read the link you showed me, and I initially tried putting the contents `CustomizeAppBuilder()` inside a try-catch block and writing the exception message to a file, but the error was not thrown there, so I didn't get anything from that step. I couldn't understand how to edit the program from the link you showed me, since I could not figure out a single method (like the usual `Main()` function) which contained the entire execution and then put it in a try-catch block.

I changed the `MainActivity` class which originally contained 
```csharp
    public class MainActivity : AvaloniaActivity<App>
    {
        protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
        {
            return base.CustomizeAppBuilder(builder);
        }
    }
```

I noticed that in the `Avalonia.Desktop` program, the `LogToTrace()` method was called to store all log information, so I figured that it would be good if I could write the log information to a file. So I changed it to this

```csharp
    public class MainActivity : AvaloniaActivity<App>
    {
        protected override AppBuilder CustomizeAppBuilder(AppBuilder builder)
        {
            string filename = GetExternalFilesDir(string.Empty) + "/tracelog.txt";
            System.Diagnostics.Trace.Listeners.Add
                (
                new System.Diagnostics.TextWriterTraceListener(filename) 
                { 
                    Filter = new System.Diagnostics.EventTypeFilter(System.Diagnostics.SourceLevels.All),
                    TraceOutputOptions = (System.Diagnostics.TraceOptions)63
                }
                );
            System.Diagnostics.Trace.Listeners.Remove("Default");
            return base.CustomizeAppBuilder(builder).LogToTrace(Avalonia.Logging.LogEventLevel.Debug);
        }
    }
```

I am using the `LogToTrace()` method which logs all information to the sink in `System.Diagnostics`. I then added a `TraceListener` to write all the log output to a file. I followed this [guide](https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/how-to-create-and-initialize-trace-listeners)

Do let me know if I'm making a silly mistake anywhere.

**Reply by timunie** [reply]
Good point. It's just available for Desktop and also for iOS but seems like it's not for Android. Now I have no other idea. Sorry.

**Reply by timunie** [reply]
Maybe you can have a look how Xamarin does it https://stackoverflow.com/questions/48499328/how-to-catch-all-exceptions-in-try-catch-block-in-xamarin-android

Disclaimer: I am not a member of Avalonia, so someone of the core team still may have a solution.

**Reply by AzuxirenLeadGuy** [reply]
Alright. I'll try this out.

Thanks for your responses BTW. I really appreciate it. 

EDIT: This was very helpful. I am finally able to view the exception message. I have updated the title. Thank you for your help.

**Comment by danwalmsley** [comment]
@AzuxirenLeadGuy something in your ui isnt measuring correctly... you have some embedded native controls?
